File: typecast-swift/Sources/Typecast/Models/VoiceModels.swift
import Foundation

// MARK: - Voice Enums

/// Gender classification for voices
public enum GenderEnum: String, Codable, Sendable {
    case male
    case female
}

/// Age group classification for voices
public enum AgeEnum: String, Codable, Sendable {
    case child
    case teenager
    case youngAdult = "young_adult"
    case middleAge = "middle_age"
    case elder
}

/// Use case categories for voices
public enum UseCaseEnum: String, Codable, Sendable {
    case announcer = "Announcer"
    case anime = "Anime"
    case audiobook = "Audiobook"
    case conversational = "Conversational"
    case documentary = "Documentary"
    case eLearning = "E-learning"
    case rapper = "Rapper"
    case game = "Game"
    case tiktokReels = "Tiktok/Reels"
    case news = "News"
    case podcast = "Podcast"
    case voicemail = "Voicemail"
    case ads = "Ads"
}

// MARK: - Voice Response Models

/// V1 Voice response (deprecated, use VoiceV2 instead)
public struct Voice: Codable, Sendable {
    public let voiceId: String
    public let voiceName: String
    public let model: TTSModel
    public let emotions: [String]
    
    enum CodingKeys: String, CodingKey {
        case voiceId = "voice_id"
        case voiceName = "voice_name"
        case model
        case emotions
    }
}

/// Model information with supported emotions
public struct ModelInfo: Codable, Sendable {
    /// TTS model version (e.g., ssfm-v21, ssfm-v30)
    public let version: TTSModel
    /// List of supported emotions for this model
    public let emotions: [String]
}

/// V2 Voice response with enhanced metadata
public struct VoiceV2: Codable, Sendable {
    /// Unique voice identifier
    public let voiceId: String
    /// Human-readable name of the voice
    public let voiceName: String
    /// List of supported TTS models with their available emotions
    public let models: [ModelInfo]
    /// Voice gender classification
    public let gender: GenderEnum?
    /// Voice age group classification
    public let age: AgeEnum?
    /// List of use case categories this voice is suitable for
    public let useCases: [String]?
    
    enum CodingKeys: String, CodingKey {
        case voiceId = "voice_id"
        case voiceName = "voice_name"
        case models
        case gender
        case age
        case useCases = "use_cases"
    }
}

// MARK: - Voice Filter

/// Filter options for V2 voices endpoint
public struct VoicesV2Filter: Sendable {
    /// Filter by TTS model
    public var model: TTSModel?
    /// Filter by gender
    public var gender: GenderEnum?
    /// Filter by age group
    public var age: AgeEnum?
    /// Filter by use case
    public var useCases: UseCaseEnum?
    
    public init(
        model: TTSModel? = nil,
        gender: GenderEnum? = nil,
        age: AgeEnum? = nil,
        useCases: UseCaseEnum? = nil
    ) {
        self.model = model
        self.gender = gender
        self.age = age
        self.useCases = useCases
    }
    
    /// Convert to query parameters dictionary
    func toQueryParams() -> [String: String] {
        var params: [String: String] = [:]
        if let model = model {
            params["model"] = model.rawValue
        }
        if let gender = gender {
            params["gender"] = gender.rawValue
        }
        if let age = age {
            params["age"] = age.rawValue
        }
        if let useCases = useCases {
            params["use_cases"] = useCases.rawValue
        }
        return params
    }
}

File: typecast-kotlin/src/main/kotlin/com/neosapience/models/VoiceModels.kt
package com.neosapience.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * V1 Voices response (deprecated, use VoiceV2Response instead)
 */
@Serializable
data class VoicesResponse(
    @SerialName("voice_id")
    val voiceId: String,
    
    @SerialName("voice_name")
    val voiceName: String,
    
    @SerialName("model")
    val model: TTSModel,
    
    @SerialName("emotions")
    val emotions: List<String>
)

/**
 * Model information with supported emotions.
 */
@Serializable
data class ModelInfo(
    @SerialName("version")
    val version: String,
    
    @SerialName("emotions")
    val emotions: List<String>
)

/**
 * V2 Voice response with enhanced metadata.
 */
@Serializable
data class VoiceV2Response(
    @SerialName("voice_id")
    val voiceId: String,
    
    @SerialName("voice_name")
    val voiceName: String,
    
    @SerialName("models")
    val models: List<ModelInfo>,
    
    @SerialName("gender")
    val gender: GenderEnum? = null,
    
    @SerialName("age")
    val age: AgeEnum? = null,
    
    @SerialName("use_cases")
    val useCases: List<String>? = null
)

/**
 * Filter options for V2 voices endpoint.
 */
data class VoicesV2Filter(
    val model: TTSModel? = null,
    val gender: GenderEnum? = null,
    val age: AgeEnum? = null,
    val useCases: UseCaseEnum? = null
) {
    companion object {
        fun builder() = Builder()
    }
    
    class Builder {
        private var model: TTSModel? = null
        private var gender: GenderEnum? = null
        private var age: AgeEnum? = null
        private var useCases: UseCaseEnum? = null
        
        fun model(value: TTSModel?) = apply { model = value }
        fun gender(value: GenderEnum?) = apply { gender = value }
        fun age(value: AgeEnum?) = apply { age = value }
        fun useCases(value: UseCaseEnum?) = apply { useCases = value }
        
        fun build() = VoicesV2Filter(model, gender, age, useCases)
    }
}
